src/composer.ts

```diff
@@ -382,22 +372,51 @@ export class Composer<C extends Context> implements Middleware.Obj<C> {
   // only invoked with properly constrained context objects, but this cannot be
   // determined automatically.
 
+  /**
+   * Generates optional middleware based on a predicate that only operates on `ctx.update`.
+   *
+   * Example:
+   * ```ts
+   * import { Composer, Update } from 'telegraf'
+   *
+   * const predicate = (u): u is Update.MessageUpdate => 'message' in u
+   * const middleware = Composer.guard(predicate, (ctx) => {
+   *   const message = ctx.update.message
+   * })
+   * ```
+   *
+   * Note that `Composer.mount('message')` is preferred over this.
+   *
+   * @param guardFn predicate to decide whether to run the middleware based on the `ctx.update` object
+   * @param fns middleware to run if the predicate returns true
+   * @see `Composer.optional` for a more generic version of this method that allows the predicate to operate on `ctx` itself
+   */
+  static guard<C extends Context, U extends tt.Update>(
+    guardFn: (u: tt.Update) => u is U,
+    ...fns: NonemptyReadonlyArray<Middleware<GuardedContext<C, U>>>
+  ): Middleware.Fn<C> {
+    return Composer.optional<C>(
+      (ctx) => guardFn(ctx.update),
+      // @ts-expect-error see explanation above
+      ...fns
+    )
+  }
+
   /**
    * Generates middleware for handling provided update types.
    */
   static mount<C extends Context, T extends tt.UpdateType | tt.MessageSubType>(
     updateType: MaybeArray<T>,
     ...fns: NonemptyReadonlyArray<Middleware<MatchedContext<C, T>>>
   ): Middleware.Fn<C> {
-    const updateTypes = normalizeTextArguments(updateType)
+    const updateTypes = reverseRenameMessageSubtypes(
+      normalizeTextArguments(updateType)
+    )
 
-    const predicate = (ctx: C) =>
-      updateTypes.includes(ctx.updateType) ||
-      // @ts-expect-error `type` is a string and not a union of literals
-      updateTypes.some((type) => ctx.updateSubTypes.includes(type))
+    const predicate = (u: tt.Update): u is tt.Update & MountMap[T] =>
+      updateTypes.some((t) => t in u || ('message' in u && t in u.message))
```

Comment:
This is cryptic, this could be avoided by overloading `Composer.optional` to support type guards... that would also solve `@ts-expect-error` in static `Composer.guard`.

Also, `normalizeTextArguments` could probably be typed better.

Reply:
I tried to overload `Composer.optional`, but did not achieve anything neat. The same is true for `normalizeTextArguments`. The best I could come up with has so many strange types to cover the corner cases (and none of my experiments worked without explicit type casts), so that I decided to drop this. After all, it's still an internal helper method.

If you feel like you can find something beautiful there that solves the problems, even if it's just a rough idea, I'd be happy to hear about it.

Reply:
I improved the performance in 3ab84b82c67b1e7ab57c4233332969758c239a46. It also makes it more readable. If you still think that this change is cryptic, which part about it is more difficult to understand than the previous implementation?

Reply:
This is great!